dynamo: use equality guards instead of id guards for Placement/DeviceMesh

bdhirsh opened the pull request:
After talking to @anijain2305, we probably can't land this since it won't work for C++ guards. But we should still be able to do better than ID_MATCH

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124401



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @yf225

bdhirsh committed:
dynamo: use equality guards instead of id guards for Placement/DeviceMesh

[ghstack-poisoned]

bdhirsh committed:
Update on "dynamo: use equality guards instead of id guards for Placement/DeviceMesh"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

wconstab commented:
Just from reading the Pr title it sounds like equality guards would be optimal for a devicemesh, but why does c++ preclude this?

bdhirsh commented:
@wconstab yeah I think that statement was wrong- c++ guards should just call back into python to run these guards.

I originally added this PR since I thought we’d need it to unblock internal since we had a small graph where one of its inputs was a Placement, and we were recompiling a bunch. We ended up fixing enough graph breaks that the Placement didnt show up as an input anymore.

We probably still should clean up and land this PR at some point though (since it should help prevent compile from being really slow when you have a lot of graph breaks with DTensor)

bdhirsh committed:
Update on "dynamo: use equality guards instead of id guards for Placement/DeviceMesh"

After talking to anijain2305, we probably can't land this since it won't work for C++ guards. But we should still be able to do better than ID_MATCH




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng

[ghstack-poisoned]

bdhirsh committed:
Update on "dynamo: use equality guards instead of id guards for Placement/DeviceMesh"

After talking to anijain2305, we probably can't land this since it won't work for C++ guards. But we should still be able to do better than ID_MATCH




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse tianyu-l yf225

[ghstack-poisoned]

wanchaol commented on torch/_dynamo/variables/builder.py:
```diff
@@ -831,15 +831,15 @@ def build_key_value(i, k, v):
         elif DeviceMeshVariable.is_device_mesh(value):
             # TODO: see if we need to add custom guard instead of a simple ID_MATCH
-            self.install_guards(GuardBuilder.ID_MATCH)
+            self.install_guards(GuardBuilder.EQUALS_MATCH)
```
For equal match guards, I wonder if custom objects that implemented `__hash__` + `__eq__` should be leveraged in our guards logic directly (i.e. DeviceMesh and Placements implemented them)? cc @anijain2305

anijain2305 replied:
Yes, this would happen automatically. In C++, EQUALS_MATCH guard calls `PyObject_RichCompareBool` between two PyObjects*. For custom objects, it will automatically call `__eq__`.

anijain2305 commented on torch/_dynamo/guards.py:
```diff
@@ -2395,6 +2411,29 @@ def add_code_part(code_part, guard, log_only=False):
         }
 
+        if torch.distributed.is_available():
+            # Need to make sure the Placement/DeviceMesh argument imports are imported
+            # in the guard code, since the equality check we use requires running the constructor.
+            # Alternative: don't use an equality check, and manually add guards on type + inner data?
+            import torch.distributed.distributed_c10d as c10d
+            from torch.distributed._tensor.placement_types import (
+                Partial,
+                Replicate,
+                Shard,
+            )
+            from torch.distributed.device_mesh import DeviceMesh
+
+            RedOpType = c10d.ReduceOp.RedOpType
+            closure_vars.update(
+                {
+                    "Shard": Shard,
+                    "Replicate": Replicate,
+                    "Partial": Partial,
+                    "DeviceMesh": DeviceMesh,
+                    "RedOpType": RedOpType,
+                }
+            )
```
Why do we need this?

anijain2305 replied:
For equals match guard in C++, we can just store the PyObject*, so there is no need of running a constructor.

bdhirsh replied:
thanks, will update. I originally added this before C++ guards landed in the PR (although I can't remember now if I actually ran into errors with python guards that required adding this or not).

bdhirsh replied:
@anijain2305 here are the logs from running the new test with `TORCH_LOGS="+guards"`:

```
(pytorch_c) [hirsheybar@devgpu001.lla3 ~/local/c/pytorch (dtensor_object_equality_guards)]$ TORCH_LOGS="+guards" python test/distributed/_tensor/test_dtensor_compile.py -k test_dtensor_dont_recompile_on_same_placement_devicemesh
Fail to import hypothesis in common_utils, tests are not derandomized
INFO:numba.cuda.cudadrv.driver:init
DEBUG: GUARDS:
DEBUG:
TREE_GUARD_MANAGER:
+- RootGuardManager
| +- DEFAULT_DEVICE: utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:462 in init_ambient_guards
| +- GLOBAL_STATE: ___check_global_state()
| +- GuardManager: source=L['x'], accessed_by=DictGetItemGuardAccessor(x)
| | +- TENSOR_MATCH: check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=True, size=[4, 4], stride=[4, 1])  # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| | +- NO_HASATTR: hasattr(L['x'], '_dynamo_dynamic_indices') == False           # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| +- GuardManager: source=L['mesh'], accessed_by=DictGetItemGuardAccessor(mesh)
| | +- EQUALS_MATCH: L['mesh'] == DeviceMesh('cuda', [0, 1])                       # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| +- GuardManager: source=L['placement'], accessed_by=DictGetItemGuardAccessor(placement)
| | +- EQUALS_MATCH: L['placement'] == Shard(dim=1)                                # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| +- GuardManager: source=G, accessed_by=GlobalsGuardAccessor
| | +- GuardManager: source=G['DTensor'], accessed_by=DictGetItemGuardAccessor(DTensor)
| | | +- ID_MATCH: ___check_obj_id(G['DTensor'], 94634789327088)                 # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| | | +- GuardManager: source=G['DTensor'].from_local, accessed_by=GetAttrGuardAccessor(from_local)
| | | | +- ID_MATCH: ___check_obj_id(G['DTensor'].from_local, 140018919113344)     # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn

DEBUG: GUARDS:
DEBUG:
TREE_GUARD_MANAGER:
+- RootGuardManager
| +- DEFAULT_DEVICE: utils_device.CURRENT_DEVICE == None                           # _dynamo/output_graph.py:462 in init_ambient_guards
| +- GLOBAL_STATE: ___check_global_state()
| +- GuardManager: source=L['x'], accessed_by=DictGetItemGuardAccessor(x)
| | +- TENSOR_MATCH: check_tensor(L['x'], Tensor, DispatchKeySet(CPU, BackendSelect, ADInplaceOrView, AutogradCPU), torch.float32, device=None, requires_grad=True, size=[4, 4], stride=[4, 1])  # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| | +- NO_HASATTR: hasattr(L['x'], '_dynamo_dynamic_indices') == False           # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| +- GuardManager: source=L['mesh'], accessed_by=DictGetItemGuardAccessor(mesh)
| | +- EQUALS_MATCH: L['mesh'] == DeviceMesh('cuda', [0, 1])                       # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| +- GuardManager: source=L['placement'], accessed_by=DictGetItemGuardAccessor(placement)
| | +- EQUALS_MATCH: L['placement'] == Partial(sum)                                # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| +- GuardManager: source=G, accessed_by=GlobalsGuardAccessor
| | +- GuardManager: source=G['DTensor'], accessed_by=DictGetItemGuardAccessor(DTensor)
| | | +- ID_MATCH: ___check_obj_id(G['DTensor'], 94634789327088)                 # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn
| | | +- GuardManager: source=G['DTensor'].from_local, accessed_by=GetAttrGuardAccessor(from_local)
| | | | +- ID_MATCH: ___check_obj_id(G['DTensor'].from_local, 140018919113344)     # dt = DTensor.from_local(x, mesh, [placement], run_check=False)  # test/distributed/_tensor/test_dtensor_compile.py:496 in fn

frames [('total', 2), ('ok', 2)]
stats [('calls_captured', 2), ('unique_graphs', 2)]
inductor []
aot_autograd [('total', 2), ('ok', 2)]
```

bdhirsh committed:
Update on "dynamo: use equality guards instead of id guards for Placement/DeviceMesh"

After talking to anijain2305, we probably can't land this since it won't work for C++ guards. But we should still be able to do better than ID_MATCH




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse tianyu-l yf225

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
